### PR TITLE
feat: header based auth 

### PR DESCRIPTION
**Pull request by iresharma**
## Summary of change
- Adds logic required to support session management using headers instead of cookies

## Related issues
None

## Test Plan
All existing tests pass and new tests have been added for header based sessions (to be added in upcoming commits)

## Documentation changes
WIP

## Checklist for important updates
- [ ] Changelog has been updated
- [x] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
   - Along with the associated array in `lib/src/version.dart`
- [ ] Changes to the version if needed
   - In `pubspec.yaml`
   - In `lib/src/version.dart`
- [x] Had installed and ran the pre-commit hook
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [x] Add tests

### CODE REVIEW

File: testHelpers/server/index.js [hunk 1]
```diff
@@ -336,6 +336,7 @@ app.post("/auth/session/refresh", async (req, res, next) => {
     noOfTimesRefreshAttemptedDuringTest += 1;
     verifySession()(req, res, err => {
         if (err) {
+            console.log(err);
```
Comment (nkshah2) [inline testHelpers/server/index.js:339]: remove this log

File: testHelpers/server/package-lock.json [hunk 2]
```diff
@@ -41,20 +41,20 @@
       }
     },
     "node_modules/@types/express": {
-      "version": "4.17.14",
-      "resolved": "https://registry.npmjs.org/@types/express/-/express-4.17.14.tgz",
-      "integrity": "sha512-TEbt+vaPFQ+xpxFLFssxUDXj5cWCxZJjIcB7Yg0k0GMHGtgtQgpvx/MUQUeAkNbA9AAGrwkAsoeItdTgS7FMyg==",
+      "version": "4.17.17",
```
Comment (nkshah2) [inline testHelpers/server/package-lock.json:44]: reset the changes in the file

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: lib/src/supertokens-http-client.dart [hunk 3]
```diff
@@ -248,22 +278,43 @@ class Client extends http.BaseClient {
...
+  }
+
+  static String _generateCookieHeader(String? oldCookie, String? newCookie) {
+    Map<String, dynamic> existingCookieHeaderMap =
+        _generateMapFromCookie(oldCookie ?? "");
+    Map<String, dynamic> newCookiesToAddMap =
+        _generateMapFromCookie(newCookie ?? "");
+    String finalCookieHeader = "";
+    existingCookieHeaderMap.keys.forEach((element) {
+      if (newCookiesToAddMap.containsKey(element)) {
```
Comment (nkshah2) [inline lib/src/supertokens-http-client.dart:311]: Cookies inside newCookiesToAddMap that are not inside existingCookieHeaderMap are never added. I dont like this approach in general, follow these steps

1. Parse both sets of cookies
2. filter existing cookies to not include any that are present in the new cookies
3. Merge the two sets of cookies
4. Create a string

File: testHelpers/server/package-lock.json (hunk 2 above)
Comment (nkshah2) [reply testHelpers/server/package-lock.json:44]: @iresharma This is still not done

File: lib/src/dio-interceptor-wrapper.dart [hunk 4]
```diff
@@ -168,19 +178,67 @@ class SuperTokensInterceptorWrapper extends Interceptor {
...
+    if (refreshHeader != null) {
+      Utils.setToken(TokenType.REFRESH, refreshHeader);
+    }
+  }
+
+  Future<RequestOptions> _removeAuthHeaderIfMatchesLocalToken(
+      RequestOptions req) async {
+    if (req.headers.containsKey("Authorization") ||
+        req.headers.containsKey("authorization")) {
+      String authValue = req.headers['Authorization'];
```
Comment (nkshah2) [inline lib/src/dio-interceptor-wrapper.dart:217]: You are checking if it contains either upper case or lower case but you are only fetching upper case inside the if block

File: lib/src/dio-interceptor-wrapper.dart [hunk 5]
```diff
@@ -168,19 +178,67 @@ class SuperTokensInterceptorWrapper extends Interceptor {
...
+  }
+
+  static Future<RequestOptions> _setAuthorizationHeaderIfRequired(
+      RequestOptions options,
+      {bool addRefreshToken = false}) async {
+    String? accessToken = await Utils.getTokenForHeaderAuth(TokenType.ACCESS);
+    String? refreshToken = await Utils.getTokenForHeaderAuth(TokenType.REFRESH);
+
+    if (accessToken != null && refreshToken != null) {
+      if (options.headers["Authorization"] != null) {
```
Comment (nkshah2) [inline lib/src/dio-interceptor-wrapper.dart:235]: Also check for lower case

File: lib/src/front-token.dart [hunk 6]
```diff
@@ -46,13 +47,19 @@ class FrontToken {
             await SuperTokensUtils.getLocalSessionState();
 
         if (localSessionState.status == LocalSessionStateStatus.EXISTS) {
-          _frontTokenMutex.acquire();
+          await _frontTokenMutex.acquire();
         } else {
           finalReturnValue = null;
+          if (_frontTokenMutex.isLocked) {
+            _frontTokenMutex.release();
```
Comment (nkshah2) [inline lib/src/front-token.dart:54]: is release not async?

File: lib/src/supertokens-http-client.dart [hunk 7]
```diff
@@ -25,6 +25,7 @@ class Client extends http.BaseClient {
   }
 
   http.Client _innerClient = http.Client();
+  http.BaseRequest? _requestForRetry;
```
Comment (nkshah2) [inline lib/src/supertokens-http-client.dart:28]: Make sure relevant comments in the dio interceptor are also fixed in this file

File: lib/src/utilities.dart [hunk 8]
```diff
@@ -103,6 +131,9 @@ class SuperTokensUtils {
 
   static Future<LocalSessionState> getLocalSessionState() async {
     var lastAccessTokenUpdate = await getFromStorage(lastAccessTokenStorageKey);
+    // ! DO NOT REMOVE THE BELOW FOR EACH LOOP, REMOVING THIS BREAKS THRE TESTS
+    var instance = await SharedPreferences.getInstance();
+    instance.getKeys().forEach((element) {});
```
Comment (nkshah2) [inline lib/src/utilities.dart:136]: Is this still relevant? Try running tests without this because this is a very bad thing to add

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: lib/src/front-token.dart (hunk 6 above)
Comment (iresharma) [reply lib/src/front-token.dart:54]: nope

File: lib/src/supertokens.dart [hunk 9]
```diff
@@ -22,15 +22,16 @@ class SuperTokens {
...
 
   static void init({
     required String apiDomain,
     String? apiBasePath,
     int sessionExpiredStatusCode = 401,
-    String? cookieDomain,
-    String? userDefaultdSuiteName,
+    String? sessionTokenBackendDomain,
+    SuperTokensTokenTransferMethod? tokenTransferMethod,
+    String? userDefaultSuiteName,
```
[the remark this answers is not in the document]
Comment (iresharma) [reply lib/src/supertokens.dart:34]: No, I'll remove it

File: lib/src/utilities.dart [hunk 10]
```diff
@@ -237,32 +265,110 @@ class Utils {
...
 
 class NormalisedInputType {
   late String apiDomain;
   late String? apiBasePath;
   late int sessionExpiredStatusCode = 401;
-  late String? cookieDomain;
-  late String? userDefaultdSuiteName;
+  late String? sessionTokenBackendDomain;
+  late SuperTokensTokenTransferMethod tokenTransferMethod;
+  late String? userDefaultSuiteName;
```
Comment (nkshah2) [inline lib/src/utilities.dart:352]: Is this used anywhere?

File: testHelpers/server/package-lock.json (hunk 2 above)
Comment (nkshah2) [reply testHelpers/server/package-lock.json]: I am still seing a diff in this file

File: lib/src/supertokens-http-client.dart [hunk 11]
```diff
@@ -248,22 +281,32 @@ class Client extends http.BaseClient {
...
+      return oldCookie;
     }
+    List<Cookie> oldCookies =
+        SuperTokensCookieStore.getCookieListFromHeader(oldCookie);
+    List<Cookie> newCookies =
+        SuperTokensCookieStore.getCookieListFromHeader(newCookie);
+    Iterable newCookiesNames = newCookies.map((e) => e.name);
+    oldCookies.removeWhere((element) => newCookiesNames.contains(element.name));
+    newCookies.addAll(oldCookies);
+    return newCookies.map((e) => e.toString()).join(':');
```
Comment (nkshah2) [inline lib/src/supertokens-http-client.dart:309]: Why are we joining with `:`?

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: lib/src/utilities.dart (hunk 10 above)
Comment (iresharma) [reply lib/src/utilities.dart:352]: nope, had copied over from iOS
Somehow missed it in previous PRs

**Review by nkshah2 — APPROVED** [review]

**Pull request merged**